File: Prefix Sum/leetcode_410.py

# Problem link ===>> https://leetcode.com/problems/split-array-largest-sum/description/?envType=problem-list-v2&envId=21r1mgbr
# submission link == =>> https://leetcode.com/problems/split-array-largest-sum/submissions/1851139313/?envType=problem-list-v2&envId=21r1mgbr

import sys
import math
import bisect
import heapq
from collections import defaultdict, deque, OrderedDict, Counter
from functools import lru_cache, reduce
from itertools import permutations, combinations, accumulate
from typing import List, Tuple, Dict, Set, Optional, Union, Any

MOD = 10**9 + 7
INF = float('inf')
input = sys.stdin.read
print = sys.stdout.write


class Solution:
    def helper(self, nums: List[int], perK: int, k: int) -> bool:
        count = 1
        sum = 0
        for num in nums:
            if (sum+num > perK):
                count += 1
                sum = num
            else:
                sum += num

        return count <= k

    def splitArray(self, nums: List[int], k: int) -> int:
        high = sum(nums)
        low = max(nums)
        soln = 0
        while (low <= high):
            mid = low + (high-low)//2
            if (self.helper(nums, mid, k)):
                soln = mid
                high = mid-1
            else:
                low = mid + 1

        return soln

def main():
    t = int(input())
    for _ in range(t):
        n = int(input())

if __name__ == "__main__":
    solution = Solution()

File: Prefix Sum/leetcode_410.cpp

// Problem link ===>> https://leetcode.com/problems/split-array-largest-sum/description/?envType=problem-list-v2&envId=21r1mgbr

#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace std;
using namespace __gnu_pbds;
#define ll long long
#define ld long double
#define endl '\n'
const int MOD = 1e9 + 7;
const double EPS = 1e-9;
const double PI = acos(-1.0);
template <typename T>
using ordered_set = tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;
template <typename K, typename V>
using ordered_map = tree<
    K, V, less<K>, rb_tree_tag, tree_order_statistics_node_update>;

class Solution
{
public:
    bool helper(vector<int> &nums, int perK, int k)
    {
        int count = 1;
        long long sum = 0;

        for (int num : nums)
        {
            if (sum + num > perK)
            {
                count++;
                sum = num;
            }
            else
            {
                sum += num;
            }
        }
        return count <= k;
    }

    int splitArray(vector<int> &nums, int k)
    {
        long long low = *max_element(nums.begin(), nums.end());
        long long high = accumulate(nums.begin(), nums.end(), 0LL);

        int soln = 0;

        while (low <= high)
        {
            long long mid = low + (high - low) / 2;

            if (helper(nums, mid, k))
            {
                soln = mid;
                high = mid - 1;
            }
            else
            {
                low = mid + 1;
            }
        }

        return soln;
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);
    
    ll t;
    cin >> t;
    while (t--)
    {
        ll n;
        cin >> n;

        
    }
    return 0;
}